### ISSUE TITLE
Things to do before merging device-rewrite

### ISSUE THREAD

**Issue by haikarainen** [haikarainen/light#62]
What needs to be done before a merge:

* ~General code cleanup, whitespace, consistent codestyle,~ const correctness, etc 
* ~Remove any GNU-specifics where possible (`strdupa` for example)~ (kinda fixed, but we use gnu extensions so were screwed anyway)
* ~Update README.md to reflect new usage~
* ~Create DOCUMENTATION.md to document codebase (how to implement new/custom enumerators etc.)~
* ~Replace any `strcmp` with `strncmp`~
* ~Create function `bool _light_get_target_path(light_context_t*, char* output_path, size_t output_size);` for better code reusal in commands etc.~
* ~closedir!~

More?

**Comment by troglobit** [comment]
Yup, lots of cleanup to be done before merging #61, but otherwise good work on this! You need to lay down the law on your preferred coding style, for instance.  I willfully just went ahead and reindented to Linux style.

What's more critical though is stuff like this:

```
$ ./src/light -h
*** buffer overflow detected ***: ./src/light terminated
Aborted (core dumped)
```

Hmmmm ... let's ask valgrind what's up :-)

```
$ valgrind ./src/light -h
[snip]
==10091== Invalid write of size 8
==10091==    at 0x10A10E: light_create_enumerator (light.c:423)
==10091==    by 0x10ADF7: light_initialize (light.c:363)
==10091==    by 0x108E75: main (main.c:13)
==10091==  Address 0x522d310 is 192 bytes inside an unallocated block of size 4,193,680 in arena "client"
[snip]
==10091== Process terminating with default action of signal 6 (SIGABRT)
==10091==    at 0x4E7AE97: raise (raise.c:51)
==10091==    by 0x4E7C800: abort (abort.c:79)
==10091==    by 0x4EC5896: __libc_message (libc_fatal.c:181)
==10091==    by 0x4F70CFE: __fortify_fail_abort (fortify_fail.c:33)
==10091==    by 0x4F70D20: __fortify_fail (fortify_fail.c:44)
==10091==    by 0x4F6EA0F: __chk_fail (chk_fail.c:28)
==10091==    by 0x4F6E1C4: __vsnprintf_chk (vsnprintf_chk.c:37)
==10091==    by 0x4F6E094: __snprintf_chk (snprintf_chk.c:34)
==10091==    by 0x10A160: snprintf (stdio2.h:64)
==10091==    by 0x10A160: light_create_enumerator (light.c:427)
==10091==    by 0x10ADF7: light_initialize (light.c:363)
==10091==    by 0x108E75: main (main.c:13)
[snip]
```
Ah, we seem to trigger the glibc fortification handling.  Maybe you can have a look at it right away since you're the one most involved in the rewrite?

**Comment by haikarainen** [comment]
Thanks! I'll get on that as soon as I find time.

**Comment by haikarainen** [comment]
@troglobit I cant recreate your crash. Please do a full make distclean/clean and rebuild?

**Comment by troglobit** [comment]
A distclean doesn't help, but if I find a way to fix it I'll post a PR instead.

**Comment by haikarainen** [comment]
What system are you running? Could you create a separate issue for this crash?

**Comment by chrboe** [comment]
https://github.com/haikarainen/light/blob/master/src/light.c#L438

Idk if it's me, but that `malloc` seems messed up. Youre allocating `sizeof(light_device_enumerator_t*)` (i.e. 8) bytes, and then writing to a member of that struct. Where is that memory coming from? I think it needs to be `malloc(sizeof(light_device_enumerator_t))`.

**Comment by chrboe** [comment]
On a side note, why does it even do the enumerator shenanigans when it just displays the help message either way?

**Comment by haikarainen** [comment]
Youre correct, thats a malplaced *.  Weird that it even runs for me!

The application workflow has been streamlined so that the enumerators are always initialized before a command is run. Doing exceptions for some commands would create more code complexity for practically nothing!

**Comment by haikarainen** [comment]
@chrboe We both seem to be working on the same thing :D

**Comment by haikarainen** [comment]
Are you doing your PR or should I continue?

**Comment by chrboe** [comment]
I'm almost finished, so I just need to create the PR

**Comment by chrboe** [comment]
>The application workflow has been streamlined so that the enumerators are always initialized before a command is run.

This is actually becoming a problem now. On my test machine, there is no backlight controller, (i.e. nothing in `/sys/class/backlight`). This means that the loop at [impl/sysfs.c:85](https://github.com/haikarainen/light/blob/master/src/impl/sysfs.c#L85) doesn't run, which means that `best_controller` is uninitialized (i.e. not a valid string).

I tried to go the obvious route and do a check for whether or not the loop found anything and return `false` if it didn't. This does work, but now the help message doesn't even show because the error aborts the program first. What to do?

**Comment by haikarainen** [comment]
@chrboe You should simply not create the auto devicetarget if no backlight targets were found, and still return true. This is a(-nother) error on my part though. An enumerator finding no targets is not an error, its simply a response to the hardware.

Either way, I think you should create one PR for the memory fixes (its pretty urgent) and then maybe focus on fixing this later (or I can do it). Did you catch the uninitialized num_devices/num_targets variables when fixing the memory errors?

**Comment by haikarainen** [comment]
Actually, if you could not do the PR and leave it to me, that would be great. I just realized its all tied up into the custom enumerator implementation documentation/usage, which means I need to further abstract out some functions or find another solution. Thanks a ton for the contributions!

**Comment by chrboe** [comment]
Lol seems like I read this a little too late, I just created the PR. It's just a few lines so you might be able to start from there anyways. If not feel free to close it

**Comment by haikarainen** [comment]
Thanks!
